components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfSignUpHandler.java

```diff
@@ -0,0 +1,232 @@
+
+                    // Lock user account.
+                    try {
+
+                        Claim claim = new Claim(IdentityMgtConstants.CLAIM_ROOT_DIALECT, IdentityMgtConstants
+                                .ACCOUNT_LOCKED_CLAIM_URI, Boolean.toString(true));
+
+                        List<Claim> claims = new ArrayList<>(1);
+                        claims.add(claim);
+
+                        identityStore.updateUserClaims(user.getUniqueUserId(), claims, null);
+                    } catch (IdentityStoreException e) {
+                        throw Utils.handleServerException(ErrorCodes.FAILED_ACCOUNT_LOCK, user
+                                .getUniqueUserId(), e);
+                    }
+
+                    UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+                    userRecoveryDataStore.invalidateByUserUniqueId(user.getUniqueUserId());
+
+                    String secretKey = Utils.generateUUID();
+                    UserRecoveryData recoveryDataDO = new UserRecoveryData(user.getUniqueUserId(), secretKey,
+                                                                           RecoveryScenarios.SELF_SIGN_UP,
+                                                                           RecoverySteps.CONFIRM_SIGN_UP);
+                    userRecoveryDataStore.store(recoveryDataDO);
+
+                    if (config.isNotificationInternallyManaged()) {
+
+                        Property[] arbitraryProperties = (Property[]) eventProperties.get(SELF_SIGN_UP_PROPERTIES);
+                        Utils.triggerNotification(eventService, user.getUniqueUserId(),
+                                                  NOTIFICATION_TYPE_ACCOUNT_CONFIRM,
+                                                  secretKey,
+                                                arbitraryProperties);
+                    } else {
+                        eventContext.addParameter(IdentityStoreConstants.UNIQUE_USED_ID, user.getUniqueUserId());
+                        eventContext.addParameter(CONFIRMATION_CODE, secretKey);
+                    }
+                } else {
+
+                    if (log.isDebugEnabled()) {
+                        log.debug("Changing user lifecycle state to {} for user: {}", UserState
+                                .UNLOCKED__UNVERIFIED.toString(), user.getUniqueUserId());
+                    }
+                    LifecycleOperationManager.executeLifecycleEvent(UserState.UNLOCKED__UNVERIFIED.toString(),
+                                                                    domainUID, domainUID, null);
+                    identityStore.setUserState(user.getUniqueUserId(), UserState.UNLOCKED__UNVERIFIED.toString());
+                }
+            } catch (LifecycleException e) {
+                throw Utils.handleServerException(ErrorCodes.FAILED_LIFECYCLE_EVENT, null, e);
```

Comment:
None of these exceptions carry any contextual information. e.g. User, what lifecycle state to which, etc.  Need to improve the Utils.handleServerException(..) if necessary.